### PR TITLE
feat: saved favorite device across sessions [WPB-783]

### PR DESCRIPTION
**Pull request by tlebon**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-783" title="WPB-783" target="_blank">StoryWPB-783</a>  [Web] Use the last used audio/microphone/camera settings of the user by default
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues
Saves users favorite devices across different setups. 

### Causes (Optional)

saving devices previously based on deviceId (which can become mapped to default and lose the actual ID)
only one setting for saved device, not a list of favorites. 

### Solutions

ranking favorites based on last used, with default with no extra devices plugged in set to last option in faves. 
also setting device list to use value not default value so list will refresh.

### CODE REVIEW

File: src/script/media/MediaDevicesHandler.ts [hunk 1]
```diff
@@ -165,15 +199,33 @@ export class MediaDevicesHandler {
    */
   private subscribeToObservables(): void {
     this.currentDeviceId.audioInput.subscribe(mediaDeviceId => {
+      const faveAudioInput = loadValue<string[]>(FavoriteDeviceTypes.AUDIO_INPUT) ?? [];
+      const newFaveAudioInput =
```
Comment (arjita-mitra) [inline src/script/media/MediaDevicesHandler.ts]: The structure for faveAudioInput, faveAudioOutput, and faveVideoInput is repetitive. We can create a reusable function to achieve the same result.

```
const getNewFavorites = (deviceType, mediaDeviceId) => {
  const favorites = loadValue<string[]>(deviceType) ?? [];
  const previousDeviceSupportCount = this.previousDeviceSupport[deviceType];
  const availableDevicesCount = this.availableDevices[deviceType]().length;

  if (previousDeviceSupportCount >= availableDevicesCount) {
    return [...favorites.filter(id => id !== mediaDeviceId), mediaDeviceId];
  }
  return [mediaDeviceId, ...favorites.filter(id => id !== mediaDeviceId)];
};

```

Comment (arjita-mitra) [reply src/script/media/MediaDevicesHandler.ts]: const newFaveAudioInput = getNewFavorites(FavoriteDeviceTypes.AUDIO_INPUT, mediaDeviceId);
const newFaveAudioOutput = getNewFavorites(FavoriteDeviceTypes.AUDIO_OUTPUT, mediaDeviceId);
const newFaveVideoInput = getNewFavorites(FavoriteDeviceTypes.VIDEO_INPUT, mediaDeviceId);

File: src/script/media/MediaDevicesHandler.ts [hunk 2]
```diff
@@ -221,8 +273,18 @@ export class MediaDevicesHandler {
...
-        currentId(devices[0]?.deviceId);
+
+      const favorites = loadValue<string[]>(`${type}-fave`) ?? [];
+      const isFavorite = favorites.map(favorite => devices.some(d => d.groupId === favorite));
+
+      if (isFavorite.some(favorite => favorite)) {
+        const faveIndex = isFavorite.findIndex(favorite => favorite);
+        currentId(
+          devices.find(d => d.deviceId === favorites[faveIndex] || d.groupId === favorites[faveIndex])?.groupId ??
+            devices[0]?.groupId,
```
Comment (arjita-mitra) [inline src/script/media/MediaDevicesHandler.ts]: devices[0]?.groupId has multiple occurrence. Using destructuring we can make it cleaner and more readable.
 
```
const [firstDevice = {}] = devices; 
const { groupId: firstDeviceGroupId } = firstDevice;
```

**Review by arjita-mitra — CHANGES_REQUESTED** [review]

File: src/script/media/MediaDevicesHandler.ts (hunk 2 above)
Comment (tlebon) [reply src/script/media/MediaDevicesHandler.ts]: i changed this slightly differently. this var is only used twice so i dont think this type of destructuring makes it more readable (to me)

File: src/script/media/MediaDevicesHandler.ts [hunk 3]
```diff
@@ -165,18 +189,37 @@ export class MediaDevicesHandler {
...
     this.currentDeviceId.audioInput.subscribe(mediaDeviceId => {
+      const faveAudioInput = loadValue<string[]>(FavoriteDeviceTypes.AUDIO_INPUT) ?? [];
+      const disconnected = this.previousDeviceSupport.audioInput > this.availableDevices.audioInput().length;
+      const newFaveAudioInput = this.orderFavoriteDevices(mediaDeviceId, faveAudioInput, disconnected);
+
       storeValue(MediaDeviceType.AUDIO_INPUT, mediaDeviceId);
+      storeValue(FavoriteDeviceTypes.AUDIO_INPUT, newFaveAudioInput);
     });
 
     this.currentDeviceId.audioOutput.subscribe(mediaDeviceId => {
```
Comment (arjita-mitra) [inline src/script/media/MediaDevicesHandler.ts]: In my opinion its good to create a reusable subscription function. If you ever need to make a change to the subscription logic, you only need to do it in one place. This reduces the risk of bugs and it's clear what each subscription is doing without having to repeat the logic. If you ever need to add more device types in the future, that would be easy too.

```
private subscribeToDevice(deviceType: 'audioInput' | 'audioOutput' | 'videoInput') {
    this.currentDeviceId[deviceType].subscribe(mediaDeviceId => {
        const faveDevices = loadValue<string[]>(FavoriteDeviceTypes[deviceType.toUpperCase()]) ?? [];
        const disconnected = this.previousDeviceSupport[deviceType] > this.availableDevices[deviceType]().length;
        const newFaveDevices = this.orderFavoriteDevices(mediaDeviceId, faveDevices, disconnected);

        storeValue(MediaDeviceType[deviceType.toUpperCase()], mediaDeviceId);
        storeValue(FavoriteDeviceTypes[deviceType.toUpperCase()], newFaveDevices);
    });
}

subscribeToDevice('audioInput');
subscribeToDevice('audioOutput');
subscribeToDevice('videoInput');
```

**Review by arjita-mitra — CHANGES_REQUESTED** [review]

File: src/script/media/MediaDevicesHandler.ts [hunk 4]
```diff
@@ -111,16 +118,27 @@ export class MediaDevicesHandler {
...
-          ((device as MediaDeviceInfo).deviceId || (device as ElectronDesktopCapturerSource).id) === currentDeviceId,
-      );
-      if (isAvailable) {
+      const isAvailable = this.availableDevices[deviceType]().find(device => {
+        return (
+          ((device as MediaDeviceInfo).deviceId || (device as ElectronDesktopCapturerSource).id) === currentDeviceId
+        );
+      });
+
+      const isFavorite = loadValue<String[]>(`${deviceType}-fave`)?.find(id => id === currentDeviceId);
```
Comment (atomrc) [inline src/script/media/MediaDevicesHandler.ts]: naming feels weird here. `is*` indicates that it's a boolean, but it seems to be a device object. 

Maybe we could rename those `availableDevice` and `favoriteDevice` , WDYT?

File: test/unit_tests/media/MediaDevicesHandlerSpec.js
Comment (atomrc) [inline test/unit_tests/media/MediaDevicesHandlerSpec.js:1]: could you see if it's easy to convert to typescript and move right next to the source file?

File: src/script/media/MediaDevicesHandler.ts [hunk 5]
```diff
@@ -111,16 +118,27 @@ export class MediaDevicesHandler {
...
+
+      function isMediaDevice(
+        device: MediaDeviceInfo | ElectronDesktopCapturerSource | undefined,
+      ): device is MediaDeviceInfo {
+        return (device as MediaDeviceInfo)?.deviceId !== undefined;
+      }
```
Comment (przemvs) [inline src/script/media/MediaDevicesHandler.ts]: Move this to guards structure :)

File: src/script/media/MediaDevicesHandler.ts [hunk 6]
```diff
@@ -111,16 +118,27 @@ export class MediaDevicesHandler {
...
+
+      if (isFavorite && isAvailable) {
         return currentDeviceId;
       }
-      return MediaDevicesHandler.CONFIG.DEFAULT_DEVICE[deviceType];
+      return isMediaDevice(isAvailable)
+        ? isAvailable?.deviceId
```
Comment (przemvs) [inline src/script/media/MediaDevicesHandler.ts]: There You always have access to deviceId :)

```suggestion
        ? isAvailable.deviceId
```

Comment (przemvs) [reply src/script/media/MediaDevicesHandler.ts]: Or device can be null

File: src/script/media/MediaDevicesHandler.ts [hunk 7]
```diff
@@ -221,8 +264,19 @@ export class MediaDevicesHandler {
     const setDevices = (type: DeviceTypes, devices: MediaDeviceInfo[]): void => {
       this.availableDevices[type](devices);
       const currentId = this.currentDeviceId[type];
-      if (!devices.some(d => d.deviceId === currentId())) {
-        currentId(devices[0]?.deviceId);
+      const firstDeviceId = devices[0]?.deviceId;
+
+      const favorites = loadValue<string[]>(`${type}-fave`) ?? [];
+      const isFavorite = favorites.map(favorite => devices.some(d => d.deviceId === favorite));
```
Comment (przemvs) [inline src/script/media/MediaDevicesHandler.ts]: Hmm.. This variable name is not consistent to this method :D Because You get array :D

Comment (arjita-mitra) [reply src/script/media/MediaDevicesHandler.ts]: favoritesFoundInDevices is one option. Also don't use shorthand 'd', rename it to device. 

const favoritesFoundInDevices = favorites.map(favorite => devices.some(device => device.deviceId === favorite));

File: src/script/components/calling/FullscreenVideoCall.tsx [hunk 8]
```diff
@@ -127,31 +128,31 @@ const FullscreenVideoCall: React.FC<FullscreenVideoCallProps> = ({
...
   );
   const showSwitchCamera = availableCameras.length > 1;
 
-  const {audioInput: currentMicrophoneDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
-    DeviceTypes.AUDIO_INPUT,
+  const {audioinput: currentMicrophoneDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
+    MediaDeviceType.AUDIO_INPUT,
   ]);
-  const {audioInput} = useKoSubscribableChildren(mediaDevicesHandler.availableDevices, [DeviceTypes.AUDIO_INPUT]);
+  const {audioinput} = useKoSubscribableChildren(mediaDevicesHandler.availableDevices, [MediaDeviceType.AUDIO_INPUT]);
```
Comment (przemvs) [inline src/script/components/calling/FullscreenVideoCall.tsx]: I think You can move it to line 136, to have this in 1 place ;)

File: src/script/components/calling/FullscreenVideoCall.tsx [hunk 9]
```diff
@@ -127,31 +128,31 @@ const FullscreenVideoCall: React.FC<FullscreenVideoCallProps> = ({
...
-      videoInput.map(device => (device as MediaDeviceInfo).deviceId || (device as ElectronDesktopCapturerSource).id),
-    [videoInput],
+      videoinput.map(device => (device as MediaDeviceInfo).deviceId || (device as ElectronDesktopCapturerSource).id),
+    [videoinput],
   );
   const showSwitchCamera = availableCameras.length > 1;
 
-  const {audioInput: currentMicrophoneDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
-    DeviceTypes.AUDIO_INPUT,
+  const {audioinput: currentMicrophoneDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
```
Comment (przemvs) [inline src/script/components/calling/FullscreenVideoCall.tsx]: I think You can move it to line 131, to have this in 1 place ;)

File: src/script/components/calling/FullscreenVideoCall.tsx [hunk 10]
```diff
@@ -127,31 +128,31 @@ const FullscreenVideoCall: React.FC<FullscreenVideoCallProps> = ({
     'classifiedDomains',
   ]);
 
-  const {videoInput: currentCameraDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
-    DeviceTypes.VIDEO_INPUT,
+  const {videoinput: currentCameraDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
```
Comment (przemvs) [inline src/script/components/calling/FullscreenVideoCall.tsx]: This doesn't work ? 🤔 

```suggestion
  const {[MediaDeviceType.VIDEO_INPUT]: currentCameraDevice} = useKoSubscribableChildren(mediaDevicesHandler.currentDeviceId, [
```

**Review by atomrc — APPROVED** [review]
LGTM

**Pull request merged**